Add Microsoft.AspNetCore.BuildTools.ApiCheck to builds

dougbu opened the pull request:
- aspnet/KoreBuild#143

Add Microsoft.AspNetCore.BuildTools.ApiCheck.Task project
  - include task assembly in the Microsoft.AspNetMicrosoft.AspNetCore.BuildTools.ApiCheck package
  - add *.props and *.targets files to hook task into `Pack` target; runs just before .nupkg is created
- reference Microsoft.AspNetMicrosoft.AspNetCore.BuildTools.ApiCheck project from the Sdk project
  - requires lots of hacks; dotnet/sdk#939 workarounds fail in this scenario

Fall back to `RuntimeTargets` when `RuntimeAssemblies` yields no assemblies

Improve project-to-project reference and native assembly handling
- avoid `FileNotFoundException` in `AssemblyLoadContext.GetAssemblyName()` due to `\placeholder\` paths
- catch and ignore `BadImageFormatException` in `AssemblyLoadContext.GetAssemblyName()`
- avoid `InvalidOperationException` in `PackageGraph.RuntimeIsCompatible()` due to missing .sha512 file

Change ApiCheck.exe's options and output
- add `api-check --compact-output` option
- generally make tool's output more useful to the task
  - catch and log `FileNotFoundException`; output types and stack traces of caught `Exception`s
- remove `--framework` option; unused in .NET Framework already
- change ApiCheck.exe option from `--ApiListing` to `--api-listing` for consistency w/ other options

nits:
- remove `PackageGraph.RuntimeIsCompatible()` `candidateRuntime` parameter
  - not needed and confusingly-named; passed value included in `compatibleRuntimes`
- update solution to use current project type GUID
- ignore global.json and launchSettings.json
- make some Linq a bit more readable
- whitespace and `using` cleanup
- move all property groups above item groups in Microsoft.AspNetMicrosoft.AspNetCore.BuildTools.ApiCheck.csproj
- make a couple of constructors `private`; used only in `public static` methods
- ignore case a bit more

notes:
- don't use `UseCommandProcessor==true`; `ToolTask` writes out .cmd file with a BOM that cmd.exe hates

dnfclas commented:
@dougbu,
Thanks for having already signed the Contribution License Agreement. Your agreement was validated by .NET Foundation. We will now review your pull request.
_Thanks,
.NET Foundation Pull Request Bot_

dougbu commented:
See also the following changes. They must go in first to avoid breaking every Universe build. Note these changes are temporary and should be replaced by `exceptions.net(45|core).json` fixes in short order. I'll file those issues once the PRs are in.
- aspnet/Caching#283
- aspnet/DataProtection#211
- aspnet/EntityFramework#7955
- aspnet/Hosting#980
- aspnet/HttpAbstractions#799
- aspnet/Localization#349
- aspnet/Microsoft.Data.Sqlite#336
- aspnet/Mvc#6013
- aspnet/Razor#1104
- aspnet/Security#1159

natemcmaster commented on src/Microsoft.AspNetCore.BuildTools.ApiCheck/build/Microsoft.AspNetCore.BuildTools.ApiCheck.targets:
```diff
@@ -0,0 +1,30 @@
+  for use outside of Microsoft.
+-->
+<Project>
+  <Target Name="ApiCheck" Condition=" '$(EnableApiCheck)' == 'true' AND $(IsPackable) == 'true' ">
```
Why does IsPackable need to be enabled for ApiCheck to run?

dougbu replied:
Because we don't check code we don't ship.

natemcmaster replied:
Pack and api check are separate and shouldn't be intermingled. It's possible to produce a binary can end up in a nuget package even when IsPackable is false. EnableApiCheck should be sufficient to disable the target.

The target can use existence of baseline.json files to avoid running. 
The target can use the `IsTestProject` property to suppress the warning on test projects.

dougbu replied:
The agreed plan was to do the API Check as part of the Pack target.

natemcmaster commented on src/Microsoft.AspNetCore.BuildTools.ApiCheck/build/Microsoft.AspNetCore.BuildTools.ApiCheck.targets:
```diff
@@ -0,0 +1,30 @@
+﻿<!--
```
This breaks one of our design patterns for tools and targets. To keep our tools composible and to separate concerns, tools packages only deliver the tool. Integration into build cycle is specified by KoreBuild and/or Internal.AspNetCore.Sdk. 

For example, the NuGetPackageVerifier package contains the tool, but the target to light it up actually is [defined in KoreBuild](https://github.com/aspnet/KoreBuild/blob/f692ca7c8ef8a277b92aef6cfc85ffdda2beafc8/build/targets/KoreBuild.SolutionBuild.targets#L344-L346).
Likewise, the `GetGitCommitInfo` task is defined and delivered by Internal.AspNetCore.BuildTools.Tasks, but the target that ties it into MSBuild compile step is [defined in Internal.AspNetCore.Sdk](https://github.com/aspnet/BuildTools/blob/e78bac3125d8baa0791205e22b884efa9224ba2f/src/Internal.AspNetCore.Sdk/build/Git.targets).

This fix should be simple. Lift the ApiCheck target into Internal.AspNetCore.Sdk instead.

dougbu replied:
Why is that odd distance between tasks and where they are used a "design pattern"?

natemcmaster replied:
https://en.wikipedia.org/wiki/Separation_of_concerns

Tasks != targets.

dougbu replied:
Done, mainly because consuming the Api Check package directly may be useful in some cases e.g. just to run the tool from a command line. Don't want to affect the consuming project then.

dougbu commented on BuildTools.sln:
```diff
@@ -15,34 +15,36 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Solution Items", "Solution
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "test", "test", "{60A938B2-D95A-403C-AA7A-3683AD64DFA0}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "NuGetPackageVerifier", "src\NuGetPackageVerifier\NuGetPackageVerifier.csproj", "{657AFF5E-164E-493D-8501-8026B7C20808}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "NuGetPackageVerifier", "src\NuGetPackageVerifier\NuGetPackageVerifier.csproj", "{657AFF5E-164E-493D-8501-8026B7C20808}"
```
🙈 Not a fan of fighting Visual Studio. We can revert this part of the change when dotnet/roslyn-project-system#1821 is fixed or wait for a future VS update to change the solution for us.

dougbu commented on src/Microsoft.AspNetCore.BuildTools.ApiCheck.Task/Microsoft.AspNetCore.BuildTools.ApiCheck.Task.csproj:
```diff
@@ -0,0 +1,14 @@
+  <Import Project="..\..\build\common.props" />
+
+  <PropertyGroup>
+    <TargetFramework>netstandard1.6</TargetFramework>
```
🙈 This is intentional. Our engineering team decided _not_ to run the tool within Visual Studio. We may change that after #201 is fixed.

dougbu commented on src/Microsoft.AspNetCore.BuildTools.ApiCheck/build/Common.props:
```diff
@@ -0,0 +1,15 @@
+<Project>
+  <!-- Set nothing if handled in project file or already done in outer build. -->
+  <PropertyGroup>
+    <EnableApiCheck Condition=" '$(EnableApiCheck)' == '' AND '$(MSBuildRuntimeType)' == 'Core' ">true</EnableApiCheck>
```
🙈 No, we don't want this in developer's faces when they are just hacking things together.

natemcmaster replied:
I disagree. We want developers aware of breaking changes. Many times their unintentionally and we don't catch them until the tool actually runs. We have developers who only build + test in VS.

But I'll let others weigh in to settle this. Should API check cause building in VS to fail? Or should it on fail on build.cmd? cc @Eilon @muratg

dougbu replied:
🙈 We've already had this conversation.

natemcmaster replied:
I wasn't aware of a design discussion about this. Please re-consider. Making VS builds and build.cmd as similar as possible is a highly desirable goal.

If the only concern is that api check errors might impede a developer's inner loop, there are better solutions. Two come to mind: (1) make api check failures a warning in Debug but an error in Release. (2) detect "BuildingInsideVisualStudio" and demote errors to being non-fatal.

Eilon replied:
It was discussed w/ the engineering team that we don't want this to run when building from VS.

dougbu commented on src/Microsoft.AspNetCore.BuildTools.ApiCheck/build/Common.props:
```diff
@@ -0,0 +1,15 @@
+    <ExcludePublicInternalTypes_InApiCheck Condition=" '$(ExcludePublicInternalTypes_InApiCheck)' == '' ">true</ExcludePublicInternalTypes_InApiCheck>
+
+    <!-- Hook API checks into Pack run, prior to creating the .nupkg but after build (if any). -->
+    <GenerateNuspecDependsOn Condition=" ! $(GenerateNuspecDependsOn.Contains('ApiCheck')) ">$(GenerateNuspecDependsOn);ApiCheck</GenerateNuspecDependsOn>
```
Sure, I'll remove the `Condition`.

Don't understand your point about `$(BuildDependsOn)` versus `$(GenerateNuspecDependsOn)`. Failures in this target cause command-line builds to fail. Also, the intent is to check the APIs of assemblies we ship (aka pack), not everything we build.

natemcmaster replied:
Producing an and API check are orthogonal concerns. GenerateNuspecDependsOn is not guaranteed to run as a part of producing an binary.

dougbu commented on src/Microsoft.AspNetCore.BuildTools.ApiCheck/build/Microsoft.AspNetCore.BuildTools.ApiCheck.props:
```diff
@@ -0,0 +1,15 @@
+  <Import Project="$(MSBuildThisFileDirectory)Common.props"/>
+
+  <PropertyGroup>
+    <_ApiCheckTaskAssembly>$(MSBuildThisFileDirectory)..\tools\netstandard1.6\Microsoft.AspNetCore.BuildTools.ApiCheck.Task.dll</_ApiCheckTaskAssembly>
```
Intentionally not part of non-Core MSBuild. But, if I left out the condition, I'll add it.

dougbu commented on src/Microsoft.AspNetCore.BuildTools.ApiCheck/Program.cs:
```diff
@@ -59,53 +58,87 @@ public static int Main(string[] args)
+                    Console.WriteLine("  --- End of inner exceptions ---");
+                }
 
+                Console.WriteLine(e.StackTrace);
```
> This whole block can be simplified to `Console.WriteLine(e.ToString())`.

Unfortunately `FileNotFoundException`s that do not mention the filename sometimes bubble here. I will check if `ReflectionTypeLoadException` overrides `ToString()` to include loader exceptions in its output.

> consider using `Console.Error.WriteLine`

Why? MSBuild doesn't use `Console.Error` at all.

dougbu commented on src/Microsoft.AspNetCore.BuildTools.ApiCheck.Task/ApiCheckTask.cs:
```diff
@@ -0,0 +1,188 @@
+        /// <inheritdoc />
+        protected override string GenerateCommandLineCommands()
+        {
+            var arguments = string.Empty;
```
🙈 Not the Right Thing:tm: for this task. `ArgumentEscaper` doesn't handle anything but spaces (not all special `bash` and `cmd` characters) and this will run in /every/ Universe build. In addition, MSBuild properties like `$(TargetPath)` are already in reasonable shape and folder names containing double quotes are incredibly unlikely.

dougbu commented on src/Microsoft.AspNetCore.BuildTools.ApiCheck/Loader/AssemblyLoader.cs:
```diff
@@ -1,33 +1,30 @@
 #if NETCOREAPP1_1
             var lockFile = new LockFileFormat().Read(assetsJson);
-            var graph = PackageGraph.Create(lockFile, framework);
+            var graph = PackageGraph.Create(lockFile, "netcoreapp1.1");
```
As we discussed offline, moving to `netcoreapp2.0` hit some roadblocks. Would likely need to build this tool against a specific shared library version to ensure consuming projects run it in something newer. For now, leaving this as-is because it's earlier than what KoreBuild installs.

On the other hand, I have preserved the dependency version cleanup I did to get rid of package downgrades after moving every `.exe` to `netcoreapp2.0`.

natemcmaster replied:
Let me clarify: I didn't mean we upgrade the tool to _run_ on netcoreapp2.0. Rather I was asking if we can allows this tool to parse the graph for multiple versions of netcoreapp1.0. 

More context: some other tools hardcode netcoreapp1.0 which led to #121. Hardcoding netcoreapp1.1 is repeating the same mistake. A fix for #121 will likely revert the removal of framework if you commit this as is.

dougbu replied:
Thanks for the context. I'll restore the `--framework` command-line option.

dougbu commented on src/Microsoft.AspNetCore.BuildTools.ApiCheck/Microsoft.AspNetCore.BuildTools.ApiCheck.csproj:
```diff
@@ -1,40 +1,62 @@
   <Import Project="..\..\build\common.props" />
+
   <PropertyGroup>
     <VersionPrefix>1.0.1</VersionPrefix>
```
Agreed. Will talk to @Eilon offline about the Sdk version. That one is much more difficult to change.

dougbu commented:
:up::date: with numerous changes. Have a look at recent commit titles to get some sense of what's changed.

This should be ready to go. Running another local Universe build to confirm. But, that's been hanging while cloning repos (before these tools are used) for some reason&hellip;

dougbu commented:
&hellip; hanging was a temporary issue with my machine. This PR is working.

natemcmaster commented on src/Internal.AspNetCore.BuildTools.Tasks/Internal.AspNetCore.BuildTools.Tasks.csproj:
```diff
@@ -18,7 +18,7 @@
 
   <ItemGroup>
-    <PackageReference Include="NETStandard.Library" Version="1.6.0" PrivateAssets="All" />
+    <PackageReference Include="NETStandard.Library" Version="$(NetStandardImplicitPackageVersion)" PrivateAssets="All" />
```
Please revert the version bumps. These versions were picked specifically to align with what is in MSBuild for .NET Core 1.0. A dependency version bump in tools packages should result in a tools version bump. This may break patch builds. We'll take care of the upgrade in #200

natemcmaster commented on src/Microsoft.AspNetCore.BuildTools.ApiCheck/Microsoft.AspNetCore.BuildTools.ApiCheck.csproj:
```diff
@@ -1,40 +1,62 @@
+  </ItemGroup>
+
+  <ItemGroup Condition=" '$(TargetFramework)' != 'netstandard1.0' ">
+    <PackageReference Include="Microsoft.Extensions.CommandLineUtils" Version="$(AspNetCoreVersion)" />
```
This package isn't being produced anymore. Use Microsoft.Extensions.CommandLineUtils.Sources.

natemcmaster commented on src/NugetReferenceResolver/NugetReferenceResolver.csproj:
```diff
@@ -9,13 +9,9 @@
-    <TargetFrameworks>netstandard1.6;net451</TargetFrameworks>
-  </PropertyGroup>
   <ItemGroup>
+    <PackageReference Include="Microsoft.DotNet.PlatformAbstractions" Version="$(AspNetCoreVersion)" />
```
Can you add a different version variable for this one? We don't produce this package and it may not be the same as AspNetCoreVersion. For example, its [latest version](https://dotnet.myget.org/feed/dotnet-core/package/nuget/Microsoft.DotNet.PlatformAbstractions) is 2.0.0-\*, not 1.2.0-\*

natemcmaster commented on src/VersionTool/VersionTool.csproj:
```diff
@@ -9,9 +9,9 @@
-    <PackageReference Include="Microsoft.DotNet.ProjectModel" Version="1.0.0-rc3-1-003177" />
+    <PackageReference Include="Microsoft.DotNet.Cli.Utils" Version="$(DotNetCliUtilsVersion)" />
+    <PackageReference Include="Microsoft.DotNet.ProjectModel" Version="$(DotNetProjectModelVersion)" />
+    <PackageReference Include="Microsoft.Extensions.CommandLineUtils" Version="$(AspNetCoreVersion)" />
```
.Sources

natemcmaster commented on src/Microsoft.AspNetCore.BuildTools.ApiCheck/Microsoft.AspNetCore.BuildTools.ApiCheck.csproj:
```diff
@@ -1,40 +1,62 @@
   <PropertyGroup>
-    <VersionPrefix>1.0.1</VersionPrefix>
-    <TargetFrameworks>netcoreapp1.1;net452</TargetFrameworks>
+    <VersionPrefix>1.0.2</VersionPrefix>
```
:+1: thanks for incrementing the version. This makes it safer to bump dependency versions.

natemcmaster commented:
Separate to the changes requested above, I wanted to re-iterate my concern about picking the concept of ApiCheck and producing a NuGet package. These are orthogonal concerns. While we _currently_ only API Check when we also pack with Microsoft.NET.Sdk, it may not _always_ be the way we want to use it. E.g. in current form, it would not work on https://github.com/aspnet/AspNetKatana or https://github.com/aspnet/EntityFramework6.

There are 2 simple changes I'm proposing to decouple API Check from pack targets:

1.  Chain off of Build instead of GenerateNuSpec
2.  Use `Condition="Exists('baseline.json')"` to determine if a API check should run, not `IsPackable=true`.

dougbu commented:
> Side note: M.Ex.CommandLineUtils doesn't exist anymore. It's now a .Sources only package cref aspnet/Common#205

That's unexpected. The .Sources package won't work for at least one project. Don't remember which one but we have tested APIs that expose `CommandLineUtils` types.

dougbu commented:
> These are orthogonal concerns.

I agree the concerns are orthogonal. However this is where we decided to hook the task in by default.

> in current form, it would not work on https://github.com/aspnet/AspNetKatana or https://github.com/aspnet/EntityFramework6.

All unusual repos will be able to hook the task in differently as they wish. The current approach works fine throughout Universe.

> not `IsPackable=true`

As it turns out, this condition is redundant. I'll remove it.

natemcmaster commented:
> where we decided

Was that decision made along with the information and concerns I have expressed? We are not bound to bad decisions.

dougbu commented:
> this bumps netcoreapp1.0 tools to 4.3.0 CoreFx packages

I agree with this part of your concern. The rest isn't valid because we have no patch branches using MSBuild. So, I'll make sure we're building for netcoreapp1.1 everywhere.

natemcmaster commented:
> The rest isn't valid because we have no patch branches using MSBuild.

Talked in person with @dougbu. I'm assuming this comment was after that because in person we discussed that we do have patch branches on MSbuild. aspnet/DotNetTools, aspnet/EntityFramework.Tools and aspnet/Scaffolding.

natemcmaster commented:
I'm dropping my requests to chain off build not pack, and to run API check in VS. I may disagree with the design, but I'm not part of the decision process so @dougbu is correct here in following the design agreed upon by the build engineering team.

dougbu commented:
:up::date: Rebased on `dev` and added 4 new commits. I restored ApiCheck.exe's `--framework` option, used that in `ApiCheckTask`, reverted many dependency version changes, remove an `$(IsPackable)` check, and used the `Microsoft.Extensions.CommandLineUtils.Sources` package where possible.

natemcmaster commented:
Nice work :)

dougbu commented on build/dependencies.props:
```diff
@@ -1,9 +1,12 @@
 ﻿<Project>
   <PropertyGroup>
     <AspNetCoreVersion>1.2.0-*</AspNetCoreVersion>
```
I'll see how much this is used now. If it's just for the `CommandLineUtils`, I'll leave it as-is. Due to patch releases, might have to use too many versions for NuGet use to be interesting and controllable from this file.

BTW I didn't change this property or add a source to the NuGet.config file 😺

dougbu commented on src/Internal.AspNetCore.Sdk/build/ApiCheck.targets:
```diff
@@ -0,0 +1,30 @@
+             Importance="normal"
+             Text="Checking for breaking changes in $(MSBuildProjectName), $(TargetFramework)" />
+    <ApiCheckTask Condition=" '$(_ApiListingFile)' != '' "
+                  ApiListingPath="$(_ApiListingFile)"
```
This is how VS indents XML by default. But, I agree it won't be easy to maintain this in Notepad and VS Code.

I am _not_ changing existing indentation to match. There's a random mix in this repo.

dougbu commented:
- d786608
- 4f67377
